src/backend/distributed/executor/insert_select_executor.c

```diff
@@ -0,0 +1,150 @@
+ * as the tuple destination.
+ */
+TupleTableSlot *
+CoordinatorInsertSelectExecScan(CustomScanState *node)
+{
+	CitusScanState *scanState = (CitusScanState *) node;
+	TupleTableSlot *resultSlot = NULL;
+
+	if (!scanState->finishedRemoteScan)
+	{
+		EState *executorState = scanState->customScanState.ss.ps.state;
+		MultiPlan *multiPlan = scanState->multiPlan;
+		Query *selectQuery = multiPlan->insertSelectQuery;
+		List *insertTargetList = multiPlan->insertTargetList;
+		Oid targetRelationId = multiPlan->targetRelationId;
+
+		ereport(DEBUG1, (errmsg("Collecting INSERT ... SELECT results on coordinator")));
+
+		ExecuteSelectIntoRelation(targetRelationId, insertTargetList, selectQuery,
+								  executorState);
+
+		scanState->finishedRemoteScan = true;
+	}
+
+	resultSlot = ReturnTupleFromTuplestore(scanState);
+
+	return resultSlot;
+}
+
+
+/*
+ * ExecuteSelectIntoRelation executes given SELECT query and inserts the
+ * results into the target relation, which is assumed to be a distributed
+ * table.
+ */
+static void
+ExecuteSelectIntoRelation(Oid targetRelationId, List *insertTargetList,
+						  Query *selectQuery, EState *executorState)
+{
+	ParamListInfo paramListInfo = executorState->es_param_list_info;
+
+	ListCell *insertTargetCell = NULL;
+	List *columnNameList = NIL;
+	bool stopOnFailure = false;
+
+	CitusCopyDestReceiver *copyDest = NULL;
+
+	BeginOrContinueCoordinatedTransaction();
+
+	/* build the list of column names for the COPY statement */
+	foreach(insertTargetCell, insertTargetList)
+	{
+		TargetEntry *insertTargetEntry = (TargetEntry *) lfirst(insertTargetCell);
+
+		columnNameList = lappend(columnNameList, insertTargetEntry->resname);
+	}
+
+	/* set up a DestReceiver that copies into the distributed table */
+	copyDest = CreateCitusCopyDestReceiver(targetRelationId, columnNameList,
+										   executorState, stopOnFailure);
+
+	ExecuteIntoDestReceiver(selectQuery, paramListInfo, (DestReceiver *) copyDest);
+
+	executorState->es_processed = copyDest->tuplesSent;
+
+	XactModificationLevel = XACT_MODIFICATION_DATA;
+}
+
+
+/*
+ * ExecuteIntoDestReceiver plans and executes a query and sends results to the given
+ * DestReceiver.
+ */
+static void
+ExecuteIntoDestReceiver(Query *query, ParamListInfo params, DestReceiver *dest)
+{
+	PlannedStmt *queryPlan = NULL;
+	Portal portal = NULL;
+	int eflags = 0;
+	int cursorOptions = 0;
+	long count = FETCH_ALL;
+
+	/* create a new portal for executing the query */
+	portal = CreateNewPortal();
+
+	/* don't display the portal in pg_cursors, it is for internal use only */
+	portal->visible = false;
+
+#if (PG_VERSION_NUM >= 90600)
+	cursorOptions = CURSOR_OPT_PARALLEL_OK;
+#endif
+
+	/* plan the subquery, this may be another distributed query */
+	queryPlan = pg_plan_query(query, cursorOptions, params);
+
+	PortalDefineQuery(portal,
+					  NULL,
+					  "",
+					  "SELECT",
+					  list_make1(queryPlan),
+					  NULL);
+
+	PortalStart(portal, params, eflags, GetActiveSnapshot());
```

Comment:
That's an annoying bug. This is caused by the way in which Citus handles calls to `nextval` in `SELECT` queries. Arguably those should not be pushed down since the sequence might not even exist on the worker. This seems like something we should fix independently of the INSERT...SELECT implementation. For now I guess we could error out if there is a `nextval` call in the insert target list and the SELECT needs distributed planning.